castai/resource_allocation_group.go

```diff
@@ -0,0 +1,303 @@
+				Description: `Operator with which to connect the labels
+	OR (default) - workload needs to have at least one label to be included
+	AND - workload needs to have all the labels to be included`,
+				Optional: true,
+				Default:  sdk.OR,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{
+					string(sdk.AND), string(sdk.OR),
+				}, false)),
+			},
+		},
+	}
+}
+
+func resourceAllocationGroupDiff(_ context.Context, d *schema.ResourceDiff, _ any) error {
+	var clusterIds []string
+	if cids, ok := d.GetOk("cluster_ids"); ok {
+		clusterIds = toClusterIds(cids.(*schema.Set).List())
+	}
+	namespaces := toStringList(d.Get("namespaces").([]interface{}))
+
+	var labels []sdk.CostreportV1beta1AllocationGroupFilterLabelValue
+	if ls, ok := d.GetOk("labels"); ok {
+		labels = toLabels(ls.(map[string]interface{}))
+	}
+
+	if len(clusterIds) == 0 && len(namespaces) == 0 && len(labels) == 0 {
+		return errors.New("allocation group must specify at least one of: cluster_ids, namespaces, or labels")
+	}
+	return nil
+}
+
+func allocationGroupImporter(ctx context.Context, d *schema.ResourceData, meta any) ([]*schema.ResourceData, error) {
+	client := meta.(*ProviderConfig).api
+	if _, err := uuid.ParseUUID(d.Id()); err != nil {
+		return nil, fmt.Errorf("error parsing uuid: %w", err)
+	}
+	resp, err := client.AllocationGroupAPIGetAllocationGroupWithResponse(ctx, d.Id())
+	if err != nil {
+		return nil, fmt.Errorf("error getting allocation group: %w", err)
+	}
+
+	err = sdk.CheckOKResponse(resp, err)
+	if err != nil {
+		return nil, fmt.Errorf("error checking response: %w", err)
+	}
+
+	d.SetId(*resp.JSON200.Id)
+	return []*schema.ResourceData{d}, nil
+}
+
+func resourceAllocationGroupRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	resp, err := client.AllocationGroupAPIGetAllocationGroupWithResponse(ctx, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if !d.IsNewResource() && resp.StatusCode() == http.StatusNotFound {
+		tflog.Warn(ctx, "Allocation group not found, removing from state", map[string]any{"id": d.Id()})
+		d.SetId("")
+		return nil
+	}
+	if err := sdk.CheckOKResponse(resp, err); err != nil {
+		return diag.FromErr(err)
+	}
+
+	ag := resp.JSON200
+
+	if err := d.Set("name", ag.Name); err != nil {
+		return diag.FromErr(fmt.Errorf("setting name: %w", err))
+	}
+	if err := d.Set("cluster_ids", *ag.Filter.ClusterIds); err != nil {
+		return diag.FromErr(fmt.Errorf("setting cluster_ids: %w", err))
+	}
+	if err := d.Set("namespaces", ag.Filter.Namespaces); err != nil {
+		return diag.FromErr(fmt.Errorf("setting namespaces: %w", err))
+	}
+	if err := d.Set("labels", fromLabels(*ag.Filter.Labels)); err != nil {
+		return diag.FromErr(fmt.Errorf("setting labels: %w", err))
+	}
+	if err := d.Set("labels_operator", ag.Filter.LabelsOperator); err != nil {
+		return diag.FromErr(fmt.Errorf("setting labels_operator: %w", err))
+	}
```

Comment:
Please check examples of `mapstructure` annotation. It should do this automatically